[NEXUS-5089] Launcher improvements

adreghiciu opened the pull request:
Depends on sonatype/sisu-bundle-launcher#4 and sonatype/sisu-litmus#8

adreghiciu committed:
[NEXUS-5188] Copy keep alive threads from sisu-jsw-utils

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5188] Optionally start the keep alive thread

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5190] Drop support for RESTART command as t is JSW specific (will not be possible to implement it via installer launcher)

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5194] Use command monitor/keep alive threads from nexus bootstrap

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5198] Use nexus-test.properties for custom system properties/app context properties

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5186] Better logging in ShutdownIfNotAlive thread

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5186] Shutdown on any ConnectException

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

jdillon commented on nexus/nexus-bootstrap/pom.xml:
```diff
@@ -44,6 +44,18 @@
+      <artifactId>junit-dep</artifactId>
+      <version>4.10</version>
+      <scope>test</scope>
+    </dependency>
```
junit-dep should be picked up transitively from litmus-testsupport.  Please do not duplicate these core test deps in projects, as it makes version management a pain.

adreghiciu replied:
The problem is that Nexus main pom uses 4.8.x in DM and that results in failures.

jdillon replied:
fffff, okay.  weren't we on a path to updating nexus to use updated junit @peterlynch ?

peterlynch replied:
yes, was waiting for 2.1 to be released before potentially destabilizing a bunch of tests. I'll move this task to my short list.

adreghiciu committed:
Cleanup code by using a strategy

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Use exposed properties names from JSWConfig

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Drop obsolete method

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Directly install/use CommandMonitor

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Use command monitor threads/talker from nexus bootstrap

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5196] Make commands supported by monitor extensible and configurable

This will allow reuse of monitor in divers configurations as for example replacing KeepAliveThread

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5196] Drop KeepAliveThread

Can be replaced with monitor + ping command

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5196] Use monitor talker to send PING commands

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5196] Replace KeepAliveThread by monitor + PING command

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Monitor should shut up when he does not have something important to say (info->debug)

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Consider a command without an id (nothing received) a PING command

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Add getter for nexus directory in bundle

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Replace JSWExec with direct calls

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5091] Use env vars as fallback from system properties

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5091] Set monitor/keep alive ports as env vars

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5091] Do not set system properties for monitor/keep alive for nexus >= 2.2

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Introduce and use a log proxy

To be able to reuse threads in an non slf4j setup (Nexus < 2.1)

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Move monitor threads and friends to own package

This will avoid collisions when both bootstrap 2.1 and above is present

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Decouple monitor threads from launcher

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Implement an Nexus <= 2.1 launcher that acts as the >=2.2 Launcher

Drop dependency on sisu-jsw-utils
Use the new jaw configuration helpers methods

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Nicer logback configuration

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Keep only ITs that matter and run them against all Nexus versions (up to date)

The rest were kind of examples and did not brought any plus regarding what was tested

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5090] Shorten the paths to target/its/<index>

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Introduce index.xsl to transform index.xml

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5090] Better els and more info

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5090] Also link to injected-test.properties for non parametrized test

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Record failure messge/stacktrace and display it as tooltip

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Drop unneeded getters

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Use xic to generate jaxb classes

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
JavaDoc

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Drop method replaced by TestIndex rule

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Drop obsolete field

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Rename TestIndex to TestIndexRule

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Refactor  NexusITFileResolver into a test data rule (will move to litmus)

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Extract TestData interface

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Extract TestIndex interface

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Tidy

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Must use the rule not the face

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Move rules to package where they will be when moved to litmus

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Add TestDataRule UT

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Resolve test data also from package

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Internal TestDataRule refactoring + JavaDoc

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Fix configuration strategy resolution to also consider 2.1.x

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5199] Check that command monitor is properly installed

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Do not directly install plugins but make configuration return plugins as overlays

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
[NEXUS-5204] Replace SelectChannelConnector with BlockingChannelConnector

in order to avoid https://bugs.eclipse.org/bugs/show_bug.cgi?id=357318#c62

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Move TestIndexRule/TestDataRule to sisu-litmus

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu commented:
Pull is now ready to be reviewed/merged

jdillon commented:
I'm still not fond of the LogProxy or the "nicer" log format.  so +0.75

adreghiciu commented:
I explained to you why we have the proxy. If is not there I cannot reuse those classes in Nexus versions bellow 2.1 and will be pity to not support testing against those versions or make copies of this tasks to be used in 2.0.x versions.

jdillon commented:
Ya I still don't like it though :-P

adreghiciu committed:
JavaDoc

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Merge remote-tracking branch 'origin/master' into launcher-improvements

adreghiciu committed:
Revert log pattern

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Make log pattern configurable

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

peterlynch commented:
The new log pattern was fine with me. Going back to %date by default is surely a mistake. My head didn't hurt looking at the new log output either ;). Configurable property is OK, but that is just icing and still leaves a less than ideal default log pattern.

I appreciate Alin trying to make it better.

Anyways none of this warrants holding up this merge any longer.

+1

peterlynch commented on nexus/nexus-bootstrap/src/main/java/org/sonatype/nexus/bootstrap/monitor/ShutdownIfNotAliveThread.java:
```diff
@@ -0,0 +1,145 @@
+
+    void shutdown()
+    {
+        log.info( "Shutting down as there is no ping response on port {}", port );
```
since this is an abnormal way to shutdown, shouldn't this be warn?

peterlynch commented on nexus/nexus-launcher/src/main/java/org/sonatype/nexus/bundle/launcher/NexusBundle.java:
```diff
@@ -25,6 +25,8 @@
 
+    File getNexusDirectory();
+
     File getWorkDirectory();
```
These two methods need javadoc with examples of what you mean in the context of the real bundle.

peterlynch commented on nexus/nexus-launcher/src/main/java/org/sonatype/nexus/bundle/launcher/internal/NexusITLauncher.java:
```diff
@@ -0,0 +1,143 @@
+import org.tanukisoftware.wrapper.WrapperManager;
+
+/**
+ * TODO
```
Explain when and why you would use this class

peterlynch commented:
+1 other than javadoc required in a few places.

adreghiciu committed:
JavaDoc

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Log the fact that keep alive cannot connect/shutdowns to WARN level as that is an exceptional case

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Drop unused socket field (we use now command monitor talker to ping instead of directly using the socket)

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
JavaDoc

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Move shutdown out of pinging

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
JavaDoc

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu commented:
@peterlynch All your concerns were addressed above

adreghiciu committed:
javaDoc/format code/optimise imports/tidy

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Make possible to use a timeout while sending commands and use it in keep alive thread

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
Allow configuration of host to be pinged for keep alive thread

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>

adreghiciu committed:
JavaDoc.

Signed-off-by: Alin Dreghiciu <adreghiciu@gmail.com>
